### ISSUE TITLE
Using dict schema and marshmallow schemas

### ISSUE THREAD

**Issue by dpgaspar** [marshmallow-code/apispec#565]
Hi,

I want to ask, if it's expected to when using `MarshmallowPlugin` all added schemas must be Marshmallow schemas, for example:

``` python
from apispec import APISpec
from apispec.ext.marshmallow import MarshmallowPlugin

spec = APISpec(
    title="Gisty",
    version="1.0.0",
    openapi_version="3.0.2",
    info=dict(description="A minimal gist API"),
    plugins=[MarshmallowPlugin()],
)

spec.components.schema(
    "Gist",
    {
        "properties": {
            "id": {"type": "integer", "format": "int64"},
            "name": {"type": "string"},
        }
    },
)

spec.path(
    path="/gist/{gist_id}",
    operations=dict(
        get=dict(
            responses={"200": {"content": {"application/json": {"schema": "Gist"}}}},
            parameters = [{'in': 'path', 'schema': {'type': 'integer'}, 'name': 'pk'}, {'in': 'query', 'name': 'q', 'schema': 'Gist'}]
        )
    ),
)

print(spec.to_yaml())
```

Crashes with:

```
ValueError: 'Gist' is not a marshmallow.Schema subclass or instance and has not been registered in the marshmallow class registry.
```

Thank you!

**Comment by lafrech** [comment]
See https://github.com/marshmallow-code/apispec/pull/554#discussion_r418696526.

I was hesitant, but this issue makes me lean forward modifying the MarshmallowPlugin to pass strings without trying to resolve them as schemas.

**Comment by dpgaspar** [comment]
Thank you for the quick reply, I'm able to make it work by:

``` python
from apispec import APISpec
from apispec.ext.marshmallow import MarshmallowPlugin
from marshmallow import Schema, fields



spec = APISpec(
    title="Gisty",
    version="1.0.0",
    openapi_version="3.0.2",
    info=dict(description="A minimal gist API"),
    plugins=[MarshmallowPlugin()],
)

spec.components.schema(
    "Gist",
    {
        "properties": {
            "id": {"type": "integer", "format": "int64"},
            "name": {"type": "string"},
        }
    },
)

spec.path(
    path="/gist/{gist_id}",
    operations=dict(
        get=dict(
            responses={"200": {"content": {"application/json": {"#ref": "#components/schemas/Gist"}}}},
            parameters = [{'in': 'path', 'schema': {'type': 'integer'}, 'name': 'pk'}, {'in': 'query', 'name': 'q', '#ref': '#components/schemas/Gist'}]
        )
    ),
)

print(spec.to_yaml())
```

Is this a good way to go?

**Comment by lafrech** [comment]
Well I guess it works as a workaround but we definitely need to fix apispec for this.

I'm not sure about this part:

    {'in': 'query', 'name': 'q', '#ref': '#components/schemas/Gist'}

but I don't have the time to check right now.

**Comment by dpgaspar** [comment]
Great to know your on it!

The:
`{'in': 'query', 'name': 'q', '#ref': '#components/schemas/Gist'}`

Is somewhat specific to Flask-AppBuilder I'm using query string like this:
`api/v1/someresource/?q=<JSON or Rison>`

**Comment by Bangertm** [comment]
@lafrech I'm not sure why you view this as a problem? The main use case for defining schemas as strings is when they are passed as yaml in a docstring. That way the user just needs add the paths and any schemas referenced in the docstring get automatically added to the spec and referenced appropriately. This saves a ton of manually constructed code.

In my applications I typically use Marshmallow schemas for just about everything, but occasionally I might have a return value that I want to document for the client but that doesn't have wider significance in the system and doesn't merit creating a Marshmallow schema. In that case I typically just inline the definition or manually construct the reference. That seems perfectly reasonable to me. I'm don't really have an expectation that APISpec would translate a string into a reference object for anything other than Marshmallow schemas.

**Comment by lafrech** [comment]
> I'm don't really have an expectation that APISpec would translate a string into a reference object for anything other than Marshmallow schemas.

This is what is shown in the [quickstart](https://github.com/marshmallow-code/apispec/blob/dev/docs/quickstart.rst), although it is broken (see #550 and #554).

Also, we do that for parameter and responses already: if string and not dict, consider it is a reference and put in a dict with $ref.

The good thing with this is that we abstract the $ref layer from the client. No need to build the ref path manually. Especially since that path depends on OAS version.

It came out wrong in my comment above. I don't mean to "pass strings without trying to resolve them as schemas", but rather consider them references if and only if they don't resolve to a schema.

I understand your point. From an API perspective, it is not ideal that a string can be interpreted both as a schema name in registry and a schema reference. In practice, I don't expect name clashes to happen, though. A user would have to manually define a schema reference with a name that is a schema name as found in the registry... for another schema.

I just pushed a commit in #554 to implement this.

**Comment by lafrech** [comment]
Fixed in #554.
